wire-compiler/src/main/java/com/squareup/wire/java/TypeWriter.java

```diff
@@ -445,65 +439,117 @@ private MethodSpec messageEquals(MessageType type) {
-          addThisIfOneOf(name, "result"), addThisIfOneOf(name, "result"), nullHashValue(field));
-      return result.build();
-    }
+    result.addStatement("int tag = reader.nextTag()");
+    result.beginControlFlow("switch (tag)");
 
-    result.addStatement("int result = hashCode");
-    result.beginControlFlow("if (result == 0)");
-    boolean afterFirstAssignment = false;
-    if (!type.extensions().isEmpty()) {
-      result.addStatement("result = extensionsHashCode()");
-      afterFirstAssignment = true;
-    }
-    for (Field field : fields) {
+    for (Field field : type.fieldsAndOneOfFields()) {
+      int tag = field.tag();
       String name = sanitize(field.name());
-      name = addThisIfOneOf(name, "result");
-      if (afterFirstAssignment) {
-        result.addStatement("result = result * 37 + ($L != null ? $L.hashCode() : $L)",
-            name, name, nullHashValue(field));
+      Type.Name fieldType = field.type();
+      boolean isEnum = javaGenerator.isEnum(fieldType);
+      CodeBlock adapter = typeAdapter(fieldType);
+
+      if (field.isRepeated()) {
+        if (field.isPacked()) {
+          if (fieldType.isScalar()) {
+            result.addStatement("case $1L: builder.$2L = reader.packed(builder.$2L, $3L); break",
+                tag, name, adapter);
+          } else {
+            TypeName fieldName = javaGenerator.typeName(fieldType);
+            result.addStatement(
+                "case $1L: builder.$2L = packedEnumOrUnknown($1L, builder.$2L, reader, $3T.ADAPTER, builder); break",
+                tag, name, fieldName);
+          }
+        } else {
+          if (fieldType.isScalar()) {
+            result.addStatement(
+                "case $1L: builder.$2L = repeated(builder.$2L, reader.value($3L)); break", tag,
+                name, adapter);
+          } else {
+            TypeName fieldName = javaGenerator.typeName(fieldType);
+            if (isEnum) {
+              result.addStatement(
+                  "case $1L: builder.$2L = repeatedEnumOrUnknown($1L, builder.$2L, reader, $3T.ADAPTER, builder); break",
+                  tag, name, fieldName);
+            } else {
+              result.addStatement(
+                  "case $1L: builder.$2L = repeatedMessage(builder.$2L, reader, $3T.ADAPTER); break",
+                  tag, name, fieldName);
+            }
+          }
+        }
+      } else if (fieldType.isScalar()) {
+        result.addStatement("case $L: builder.$L = reader.value($L); break", tag, name, adapter);
       } else {
-        result.addStatement("result = $L != null ? $L.hashCode() : $L",
-            name, name, nullHashValue(field));
-        afterFirstAssignment = true;
+        TypeName fieldName = javaGenerator.typeName(fieldType);
+        if (isEnum) {
+          result.addStatement(
+              "case $1L: builder.$2L = enumOrUnknown($1L, reader, $3T.ADAPTER, builder); break",
```

Comment:
not sure I track why `unknown` and enums go together, but I suspect I'll get to that